tools/azure-sdk-tools/ci_tools/scenario/generation.py

```diff
@@ -0,0 +1,474 @@
+
+    :param str file: the absolute path to the dev_requirements.txt file
+    :param str pkg_root: the absolute path to the package's root
+    :return: None
+    """
+    adjusted_req_lines = []
+
+    with open(file, "r") as f:
+        for line in f:
+            args = [part.strip() for part in line.split() if part and not part.strip() == "-e"]
+            amended_line = " ".join(args)
+
+            if amended_line.endswith("]"):
+                trim_amount = amended_line[::-1].index("[") + 1
+                amended_line = amended_line[0 : (len(amended_line) - trim_amount)]
+
+            adjusted_req_lines.append(amended_line)
+
+    adjusted_req_lines = list(map(lambda x: build_whl_for_req(x, pkg_root), adjusted_req_lines))
+    install_deps_commands = [
+        sys.executable,
+        "-m",
+        "pip",
+        "install",
+    ]
+    logging.info(f"Installing dev requirements from freshly built packages: {adjusted_req_lines}")
+    install_deps_commands.extend(adjusted_req_lines)
+    subprocess.check_call(install_deps_commands)
+    shutil.rmtree(os.path.join(pkg_root, ".tmp_whl_dir"))
+
+
+def build_whl_for_req(req: str, package_path: str) -> str:
+    """Builds a whl from the dev_requirements file.
+
+    :param str req: a requirement from the dev_requirements.txt
+    :param str package_path: the absolute path to the package's root
+    :return: The absolute path to the whl built or the requirement if a third-party package
+    """
+
+    if ".." in req:
+        # Create temp path if it doesn't exist
+        temp_dir = os.path.join(package_path, ".tmp_whl_dir")
+        if not os.path.exists(temp_dir):
+            os.mkdir(temp_dir)
+
+        req_pkg_path = os.path.abspath(os.path.join(package_path, req.replace("\n", "")))
+        parsed = ParsedSetup.from_path(req_pkg_path)
+
+        logging.info("Building wheel for package {}".format(parsed.name))
+        create_package(req_pkg_path, temp_dir, enable_sdist=False)
+
+        whl_path = os.path.join(temp_dir, find_whl(temp_dir, parsed.name, parsed.version))
+        logging.info("Wheel for package {0} is {1}".format(parsed.name, whl_path))
+        logging.info("Replacing dev requirement. Old requirement:{0}, New requirement:{1}".format(req, whl_path))
+        return whl_path
+    else:
+        return req
+
+
+def build_and_discover_package(setuppy_path: str, dist_dir: str, target_setup: str, package_type):
+    if package_type == "wheel":
+        create_package(setuppy_path, dist_dir, enable_sdist=False)
+    else:
+        create_package(setuppy_path, dist_dir, enable_wheel=False)
+
+    prebuilt_packages = [
+        f for f in os.listdir(dist_dir) if f.endswith(".whl" if package_type == "wheel" else ".tar.gz")
+    ]
+
+    if not in_ci():
+        logging.info("Cleaning up build directories and files")
+        cleanup_build_artifacts(target_setup)
+    return prebuilt_packages
+
+
+def prepare_and_test(mapped_args: argparse.Namespace) -> int:
```

Comment:
Nit: if this only handles optional environments, we'll probably want to make the function name more indicative of that. Maybe `prepare_and_test_optional`?

Reply:
Agreed.